### ISSUE TITLE
Isabelle's DenseNet121 model is not working in gRPC.

### ISSUE THREAD

**Issue by minseok-ryu** [APPFL/APPFL#65]
**[Issue]** 
The batch normalization used in Isabelle's DenseNet121 model results in some states to have integer values. For example, ``model.state_dict()["conv1_bn.num_batches_tracked"] = 0``.

The followings are problematic:
- ``flat = np.frombuffer(response.data_bytes, dtype=np.float32)`` in ``src/appfl/protos/client.py``
- ``flat = np.frombuffer(tensor.data_bytes, dtype=np.float32)`` in ``src/appfl/protos/operator.py``
 
**[Solution]** 
Take ``dtype`` as an input. 
 
- ``src/appfl/protos/federated_learning.proto``

```python
message TensorRecord {
    string         name       = 1;
    repeated int32 data_shape = 2;
    bytes          data_bytes = 3;
    string         data_dtype = 4;
}
```
Compile:

```python
python -m grpc_tools.protoc --proto_path=. --python_out=. --grpc_python_out=. federated_learning.proto
```

- ``src/appfl/protos/utils.py`` 

```python
def construct_tensor_record(name, nparray):    
    return TensorRecord(name=name, data_shape=list(nparray.shape), data_bytes=nparray.tobytes(order='C'), data_dtype="np."+str(nparray.dtype))
```
Then, 
- ``flat = np.frombuffer(response.data_bytes, dtype=eval(response.data_dtype))        `` in ``src/appfl/protos/client.py``

**Comment by minseok-ryu** [comment]
[Issue 2]
```
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.RESOURCE_EXHAUSTED
        details = "Sent message larger than max (10485770 vs. 10485760)"
        debug_error_string = "{"created":"@1647390876.645401000","description":"Error received from peer ipv6:[::1]:50051","file":"src/core/lib/surface/call.cc","file_line":1064,"grpc_message":"Sent message larger than max (10485770 vs. 10485760)","grpc_status":8}"
```

**Comment by kibaekkim** [comment]
@minseok-ryu 
```python
        if response.data_dtype == "int64":
            flat = np.frombuffer(response.data_bytes, dtype=np.int64)
        elif response.data_dtype == "float32":
            flat = np.frombuffer(response.data_bytes, dtype=np.float32)
```
For the code above, if you had `response.data_dtype = "np.int64"` or `response.data_dtype = "np.float64"`, you could do the following:
```python
        flat = np.frombuffer(response.data_bytes, dtype=eval(response.data_dtype))
```

**Comment by minseok-ryu** [comment]
@kibaekkim 

Unfortunately, there is no ``np`` in ``response.data_dtype``. 

```python
def construct_tensor_record(name, nparray):    
    print("nparray.dtype=", nparray.dtype, " str=", str(nparray.dtype))
    return TensorRecord(name=name, data_shape=list(nparray.shape), data_bytes=nparray.tobytes(order='C'), data_dtype=str(nparray.dtype))
```

```
nparray.dtype= float32  str= float32
```

**Comment by minseok-ryu** [comment]
[Issue 3] Some issues on constructing a docker container for the ``server``. This needs to be discussed.

**Comment by minseok-ryu** [comment]
[Issue]
```
_pad(Tensor input, int[] pad, str mode="constant", float value=0.) -> (Tensor):
Expected a value of type 'float' for argument 'value' but instead found type 'int'.
:
  File "./gcloud/IsabelleTorch.py", line 272
    def forward(self, x):
        zero_padding2d_1 = F.pad(x, (3, 3, 3, 3), mode = 'constant', value = 0)
                           ~~~~~ <--- HERE
        conv1_conv      = self.conv1_conv(zero_padding2d_1)
        conv1_bn        = self.conv1_bn(conv1_conv)
"
        debug_error_string = "{"created":"@1647493171.256469000","description":"Error received from peer ipv6:[::1]:50051","file":"src/core/lib/surface/call.cc","file_line":904,"grpc_message":"Exception calling application: \n\n_pad(Tensor input, int[] pad, str mode="constant", float value=0.) -> (Tensor):\nExpected a value of type 'float' for argument 'value' but instead found type 'int'.\n:\n  File "./gcloud/IsabelleTorch.py", line 272\n    def forward(self, x):\n        zero_padding2d_1 = F.pad(x, (3, 3, 3, 3), mode = 'constant', value = 0)\n                           ~~~~~ <--- HERE\n        conv1_conv      = self.conv1_conv(zero_padding2d_1)\n        conv1_bn        = self.conv1_bn(conv1_conv)\n","grpc_status":2}"
```

```python
def save_model(model, cfg: DictConfig):
    dir = cfg.save_model_dirname
    if os.path.isdir(dir) == False:
        os.mkdir(dir)
    model_name=cfg.save_model_filename
    
    file_ext = ".pt"
    file = dir + "/%s%s" % (model_name, file_ext)
    uniq = 1
    while os.path.exists(file):
        file = dir + "/%s_%d%s" % (model_name, uniq, file_ext)
        uniq += 1
     
    model_scripted = torch.jit.script(model) # Export to TorchScript
    model_scripted.save(file) # Save
```
Isabelle's model cannot be stored by ``torch.jit.script``
